File: LeetCode-Java/Combinations.java
public class Solution {
    public List<List<Integer>> combine(int n, int k) {
        ArrayList<Integer> record = new ArrayList<Integer>();
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        if(k > n) return result;
        combinePart(n, k, 1, record, result);
        return result;
    }
    
    public void combinePart(int n, int k, int start, List<Integer> record, List<List<Integer>> result)
    {
        if( n - start + 1 < k) return;
        if(k == 0)
        {
            ArrayList<Integer> tempResult = new ArrayList<Integer>(record);
            result.add(tempResult);
        }
        else
        {
            for(int i = start; i <= n - k + 1;i ++)
            {
                record.add(i);
                int length = record.size();
                combinePart(n, k - 1, i + 1, record, result);
                record.remove(length - 1);
            }
        }
        return;
    }
}

File: Leetcode-cpp/combinations.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> result;
        vector<bool> record(n, false);
        generateCominbation(result, record, 0, k);
        return result;
    }
    void generateCominbation(vector<vector<int>>& result, vector<bool>& record, int index, int k) {
        if (k == 0) {
            vector<int> tempResult;
            for (int i = 0; i < record.size(); ++i) {
                if (record[i]) tempResult.push_back(i + 1);
            }
            result.push_back(tempResult);
        } else if (index < record.size()) {
            generateCominbation(result, record, index + 1, k);
            record[index] = true;
            generateCominbation(result, record, index + 1, k - 1);
            record[index] = false;
        }
    }
};